### PR TITLE
Runes/basic runes rework

### PR DESCRIPTION
**Pull request by Shushishtok**
I'm too lazy to actually make a PR documentation, so I'll just paste the instructions of the stuff needed:

- Completely remove wGetName(), use wGetRawName() instead.
- Default values in ProjectileOptions for caster, ability, targetteam, flags, types.
- Vision radius not necessary (no fog), should be completely removed
- ProjectilePathModifier should be reworked to have inherited classes, each with their unique properties, replace setVal to setX functions, e.g. setMaxAngle.
- Replace step_val with levelsNeeded in wGetLevelSpecialValueFor. Change current behavior to adding the inc every x levels. Defaults to 1 when omitted.
- Change rune element to be an array of elements, and change all related functions to match. Make it possible to define a single element, which still puts it in an array when the rune is created.
- Rarities: add rarity property: basic, rare, unique.
- Add function DoesUniqueRuneExist that checks if a passed rune name already exists in the game.

In a separate PR, I will rework the actual runes and their generic effects.
Fixes https://github.com/arcadia-redux/reVolt/issues/91
Fixes https://github.com/arcadia-redux/reVolt/issues/7
Fixed https://github.com/arcadia-redux/reVolt/issues/105
Fixed https://github.com/arcadia-redux/reVolt/issues/104

### CODE REVIEW

File: game/scripts/vscripts/core/path_modifier.ts [hunk 1]
```diff
@@ -1,164 +1,127 @@
...
-	//optVal can have different effects, depending on the type you use
-	setVal(name: string, value: any): void {
-		this.optVals.setVal(name, value);
+	/**
+	 * Creates a snake path modifier, which can have the amplitude and interval set.
+	 * This causes the projectile to move in a snake-like form in the direction.
+	 * The class includes methods to set its variables.
+	 * @returns A projectile path modifier class instance for snake path modifiers.
+	 */
+	static CreateSnakePathModifier() {
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/path_modifier.ts:28]: its actually SinusSlow and not SinusPath

File: game/scripts/vscripts/core/path_modifier.ts [hunk 2]
```diff
@@ -1,164 +1,127 @@
...
+		return new ProjectilePathModifierSinus();
 	}
 
-	isActive(curState: ProjectileState): boolean {
-		if (this.activeState == undefined) {
-			return curState !== ProjectileState.DEAD;
-		}
-		return curState == this.activeState;
+	/**
+	 * Creates a path modifier, which causes the projectile to change a state to the new state.
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/path_modifier.ts:43]: Maybe add a not here, that different modes are available and that these specify the function of this modifier

File: game/scripts/vscripts/core/weapon_base.ts [hunk 3]
```diff
@@ -20,6 +19,8 @@ export class BaseWeapon {
 	private maxLevel: number = 10;
 	private upgradeCosts: number[] = [];
 	private slot?: number;
+	private elements: WeaponElement[] = [];
+	runeModel?: CDOTA_BaseNPC;
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/weapon_base.ts:23]: any reason why this isn't private?

File: game/scripts/vscripts/core/weapons.ts [hunk 4]
```diff
@@ -16,7 +16,8 @@ export namespace Weapons {
 			costIncrease: 20,
 			icon: "black_gem_01.png",
 			color: WeaponColor.BLACK,
-			element: WeaponElement.CHAOS,
+			elements: [WeaponElement.CHAOS, WeaponElement.EARTH],
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/weapons.ts:19]: revert these test values please

File: game/scripts/vscripts/core/weapons.ts [hunk 5]
```diff
@@ -50,7 +51,8 @@ export namespace Weapons {
 			costIncrease: 20,
 			icon: "violet_gem_01.png",
 			color: WeaponColor.PURPLE,
-			element: WeaponElement.ELECTRICITY,
+			elements: WeaponElement.ELECTRICITY,
+			rarity: WeaponRarity.RARE,
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/weapons.ts:55]: revert these test values please

File: game/scripts/vscripts/core/weapons.ts [hunk 6]
```diff
@@ -78,81 +80,14 @@ export namespace Weapons {
...
-
 		RegisterWeapon({
 			name: WeaponName.BasicFire,
 			baseCost: 100,
 			costIncrease: 20,
 			icon: "red_gem_01.png",
 			color: WeaponColor.RED,
-			element: WeaponElement.FIRE,
+			elements: WeaponElement.FIRE,
+			rarity: WeaponRarity.UNIQUE,
```
Comment (Nibuja05) [inline game/scripts/vscripts/core/weapons.ts:90]: revert these test values please

File: game/scripts/vscripts/weapons/basic_fire.ts [hunk 7]
```diff
@@ -28,14 +24,13 @@ export class basic_fire extends BaseWeapon {
 			startLoc: loc,
 			trailEffect: "particles/projectiles/fireball/fireball.vpcf",
 			radius: radius,
-			distance: distance,
+			maxDistance: distance,
 			iUnitTargetTeam: this.GetAbilityTargetTeam(),
```
Comment (Nibuja05) [inline game/scripts/vscripts/weapons/basic_fire.ts:28]: we have default value's now, so we can remove these values here

File: game/scripts/vscripts/weapons/basic_ice.ts [hunk 8]
```diff
@@ -28,14 +24,13 @@ export class basic_ice extends BaseWeapon {
 			startLoc: loc,
 			trailEffect: "particles/projectiles/ice_shards/ice_shards.vpcf",
 			radius: radius,
-			distance: distance,
+			maxDistance: distance,
 			iUnitTargetTeam: this.GetAbilityTargetTeam(),
```
Comment (Nibuja05) [inline game/scripts/vscripts/weapons/basic_ice.ts:28]: we have default value's now, so we can remove these values here

File: game/scripts/vscripts/weapons/basic_water.ts [hunk 9]
```diff
@@ -30,14 +26,13 @@ export class basic_water extends BaseWeapon {
 			startLoc: loc,
 			trailEffect: "particles/projectiles/water_ball/water_ball.vpcf",
 			radius: radius,
-			distance: distance,
+			maxDistance: distance,
 			iUnitTargetTeam: this.GetAbilityTargetTeam(),
```
Comment (Nibuja05) [inline game/scripts/vscripts/weapons/basic_water.ts:30]: we have default value's now, so we can remove these values here

File: game/scripts/vscripts/weapons/basic_wind.ts [hunk 10]
```diff
@@ -39,14 +35,13 @@ export class basic_wind extends BaseWeapon {
 			trailEffect: "particles/projectiles/invisible_projectile.vpcf",
 			radius: radius * 3,
 			startRadius: radius,
-			distance: distance,
+			maxDistance: distance,
 			iUnitTargetTeam: this.GetAbilityTargetTeam(),
```
Comment (Nibuja05) [inline game/scripts/vscripts/weapons/basic_wind.ts:39]: we have default value's now, so we can remove these values here

**Review by Nibuja05 — COMMENTED** [review]
Good job! Everything looking fire 🔥 

Just some small mistakes here there, which should be super easy to fix ;)

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 11]
```diff
@@ -384,7 +408,16 @@ abstract class PanoramaWeaponBase {
...
+		let elementText = "";
+		for (let index = 0; index < elements.length; index++) {
+			if (index !== 0) {
+				elementText += ", ";
+			}
+
+			const element = elements[index];
+			elementText += GetElementName(element);
+		}
+		tooltipWeaponInfoElement.text = elementText;
```
[the remark this answers is not in the document]
Comment (Shushishtok) [reply content/panorama/scripts/custom_game/weapon_base.ts:420]: Removed the color completely.

[the remark this answers is not in the document]
Comment (Shushishtok) [reply content/panorama/scripts/custom_game/weapon_base.ts:420]: Also replaced the comma with a space.

File: game/scripts/vscripts/core/path_modifier.ts (hunk 2 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/path_modifier.ts:43]: Added some more notes explaining the modes.

File: game/scripts/vscripts/core/weapon_base.ts (hunk 3 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/weapon_base.ts:23]: Needed to set it in `slots.ts`. I've made SetRuneModelUnit() instead, which should only be used by `slots.ts` and nowhere else - the documentation explains that.

File: game/scripts/vscripts/core/path_modifier.ts (hunk 1 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/path_modifier.ts:28]: Fixed.

File: game/scripts/vscripts/core/weapons.ts (hunk 4 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/weapons.ts:19]: Reverted.

File: game/scripts/vscripts/core/weapons.ts (hunk 6 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/weapons.ts:90]: Reverted.

File: game/scripts/vscripts/weapons/basic_ice.ts (hunk 8 above)
Comment (Shushishtok) [reply game/scripts/vscripts/weapons/basic_ice.ts:28]: Done.

File: game/scripts/vscripts/weapons/basic_wind.ts (hunk 10 above)
Comment (Shushishtok) [reply game/scripts/vscripts/weapons/basic_wind.ts:39]: Done.

File: game/scripts/vscripts/weapons/basic_water.ts (hunk 9 above)
Comment (Shushishtok) [reply game/scripts/vscripts/weapons/basic_water.ts]: Fixed.

File: game/scripts/vscripts/core/weapons.ts (hunk 5 above)
Comment (Shushishtok) [reply game/scripts/vscripts/core/weapons.ts:55]: Reverted.

File: game/scripts/vscripts/weapons/basic_fire.ts (hunk 7 above)
Comment (Shushishtok) [reply game/scripts/vscripts/weapons/basic_fire.ts:28]: Removed.

**Review by Nibuja05 — APPROVED** [review]

**Pull request merged**